pkg/version/compare.go

```diff
@@ -113,8 +114,17 @@ func CompareVersionsOr(versionA, versionB string, incomparableRes int) int {
 	// Because we always bump release and rc version before release date, dev and nightly version is greater than
 	// release and rc version.
 	// for example: 3.0.58.x-1 > 3.0.58.0-rc.1
-	versionA = strings.Replace(versionA, "x", strconv.Itoa(math.MaxInt32), 1)
-	versionB = strings.Replace(versionB, "x", strconv.Itoa(math.MaxInt32), 1)
+	maxInt := strconv.Itoa(math.MaxInt32)
+
+	if isDevVersionIsGreaterThanRelease(versionA) && (kindA == NightlyKind || kindA == DevelopmentKind) {
+		versionA = strings.Replace(versionA, "x", maxInt, 1)
+		versionA = strings.Replace(versionA, ".0-", fmt.Sprintf(".%s-", maxInt), 1)
+	}
+	if isDevVersionIsGreaterThanRelease(versionB) && (kindB == NightlyKind || kindB == DevelopmentKind) {
+		versionB = strings.Replace(versionB, "x", maxInt, 1)
+		versionB = strings.Replace(versionB, ".0-", fmt.Sprintf(".%s-", maxInt), 1)
+	}
```

Comment:
The `strings.Replace` duplication here and in `getEffectVersion` is a sign that we do something wrong, the code structure needs an improvement so that the duplication is not present.